### ISSUE TITLE
 Test failures with bowtie2 2.3.2

### ISSUE THREAD

**Issue by satta** [sanger-pathogens/ariba#190]
Similarly to #170, we again see some differring test data with a newer version of Bowtie2:

```
======================================================================
FAIL: Test run_bowtie2 unsorted
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/1st/ariba-2.10.0+ds/ariba/tests/mapping_test.py", line 62, in test_run_bowtie2
    self.assertListEqual(expected, got)
AssertionError: Lists differ: [('1', 99, 'ref', 0, [(4, 5), (0, 20)], 'AGCCCTC[857 chars]AT')] != [('1', 153, 'ref', 30, [(0, 25)], 'AGGATACAGATCT[795 chars]AT')]

First differing element 0:
('1', 99, 'ref', 0, [(4, 5), (0, 20)], 'AGCCCTCCACAGGATGGTGGTATAC')
('1', 153, 'ref', 30, [(0, 25)], 'AGGATACAGATCTTGTGGGAAAGGT')

- [('1', 99, 'ref', 0, [(4, 5), (0, 20)], 'AGCCCTCCACAGGATGGTGGTATAC'),
-  ('1', 147, 'ref', 30, [(0, 25)], 'AGGATACAGATCTTGTGGGAAAGGT'),
? ^       ^^

+ [('1', 153, 'ref', 30, [(0, 25)], 'AGGATACAGATCTTGTGGGAAAGGT'),
? ^       ^^

+  ('1', 101, None, 30, [], 'AGCCCTCCACAGGATGGTGGTATAC'),
-  ('2', 99, 'ref', 124, [(0, 25)], 'TAATGTTCTTAGGGCTTACCATAGA'),
?        ^^

+  ('2', 73, 'ref', 124, [(0, 25)], 'TAATGTTCTTAGGGCTTACCATAGA'),
?        ^^

-  ('2', 147, 'ref', 170, [(0, 20), (4, 5)], 'TCCACCTTAGCTAAGCGCAGACTCG'),
+  ('2', 133, None, 124, [], 'CGAGTCTGCGCTTAGCTAAGGTGGA'),
   ('3', 73, 'ref', 86, [(0, 25)], 'TCGGGTCTGTACAAGGACGGATGGT'),
   ('3', 133, None, 86, [], 'CGTACTGACTGACTGACGTACTGCA'),
   ('4', 99, 'ref', 55, [(0, 25)], 'CCGCCGGGAAGTCCTTCTGTCGTGC'),
   ('4', 147, 'ref', 136, [(0, 25)], 'GGCTTACCATAGAGGTACACTAAAA'),
-  ('5', 99, 'ref', 0, [(4, 2), (0, 23)], 'CCTCCACAGGATGGTGGTATACCTG'),
?        ^^  ^^ --  ^   ---------------

+  ('5', 77, None, -1, [], 'CCTCCACAGGATGGTGGTATACCTG'),
?        ^^  ^^^   ^^

-  ('5', 147, 'ref', 166, [(0, 24), (4, 1)], 'TTCATCCACCTTAGCTAAGCGCAGA'),
+  ('5', 141, None, -1, [], 'TCTGCGCTTAGCTAAGGTGGATGAA'),
   ('6', 77, None, -1, [], 'CAGTTGCATGACGTCATGCAGTCAT'),
   ('6', 141, None, -1, [], 'AATGAGTATGATGAGTAATGGTATG'),
   ('7', 99, 'ref', 56, [(4, 1), (0, 23), (4, 1)], 'ACGCCGGGAAGTCCTTCTGTCGTGT'),
   ('7', 147, 'ref', 136, [(0, 24), (4, 1)], 'GGCTTACCATAGAGGTACACTAAAT')]

======================================================================
FAIL: Test run_bowtie2 sorted
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/1st/ariba-2.10.0+ds/ariba/tests/mapping_test.py", line 105, in test_run_bowtie2_and_sort
    self.assertListEqual(expected, got)
AssertionError: Lists differ: [('1', 99, 'ref', 0, [(4, 5), (0, 20)], 'AGCCCTC[857 chars]TG')] != [('1', 101, None, 30, [], 'AGCCCTCCACAGGATGGTGGT[795 chars]TG')]

First differing element 0:
('1', 99, 'ref', 0, [(4, 5), (0, 20)], 'AGCCCTCCACAGGATGGTGGTATAC')
('1', 101, None, 30, [], 'AGCCCTCCACAGGATGGTGGTATAC')

Diff is 1434 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 335 tests in 173.301s
```

Please also see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=870362.

**Comment by martinghunt** [comment]
I guess requiring a specific (older) version of bowtie2 is not an option? This is a pain.

**Comment by satta** [comment]
No, Debian always carries the latest packaged version, which follows upstream closely. (Also, non-Debian ariba users may have any version installed as well!)

I guess the question is whether it actually helps here to test, in essence, the output of an external tool which we have no control over... wouldn't it help to test whether the internal wrapper code does the call correctly and some correctly formed data comes back?

**Comment by martinghunt** [comment]
Except for debian, the easiest solution for me is to have ARIBA require a specific version of bowtie2, which behaves how I expect it to. ARIBA will look first at the env variable $ARIBA_BOWTIE2, so it's easy enough for ARIBA to use a different version from whatever is installed.

The thing is, it's testing whether the read mapper is doing what I need it to, not just that the format of the data is correct. The failed test is running bowtie2 and getting back unmapped reads. Those reads should map and I'd be worried about the results from ARIBA if those reads don't get mapped.

**Comment by satta** [comment]
I see. Well, as I said with most packaged versions of ariba that use dependencies to other packaged versions of bowtie2 the user would need to manually download the required version and set `ARIBA_BOWTIE2`. It might be useful to probably state this in the README because I figure that most users would intuitively want to go with the latest version even if not using the packaged ariba.

Regarding the current issue, for me the most important part here is that the tests fail, and that is currently breaking our builds, threatening ariba's existence in Debian altogether. What would you think about the following approach:
 - During the tests, check whether the version of Bowtie2 present matches the one you expect to behave like you want it to (mapping all the reads it should map).
 - If this is the case, fail the test with an error if unmapped reads occur in the test output.
 - If this is not the case, do not fail the test, but prominently display a notice message that the version of Bowtie2 present on the system is not the supported version and that you can only guarantee best results with the version you recommend (I remember that's how it was done in circlator(?) as well).

**Comment by martinghunt** [comment]
That sounds sensible. I'll also add in a big warning to STDERR at run time if the version of Bowtie2 is not the supported one.

**Comment by satta** [comment]
Great! Looking forward to a release :)

**Comment by martinghunt** [comment]
It turns out that I was checking if bowtie2 version was equal to 2.3.1 when I fixed #170. I just changed the code to check >= 2.3.1 instead. The tests now pass with Bowtie2 2.3.2, and should continue to do so unless bowtie2's behaviour gets changed again. New release is 2.10.1. Thanks again!

**Comment by satta** [comment]
Cool, many thanks. Updated in packaging now.
